Rework the bootstrap process into orchestration

NicolasT opened the pull request:
The intent of this changeset is to turn the `bootstrap.sh` script into something that invokes as little `salt` calls as possible, i.e. 'orchestrate' the whole bootstrap from Salt.

This happens in two stages:

- Bring the node into `bootstrap` state, locally. This implies running the registry, repository, `salt-master`,...
- Once `salt-master` is available, bring up the bootstrap node to a single-node control-plane through orchestration, which is *mostly*

  * Sync the master
  * Bring the bootstrap minion to a fabricated highstate
  * Deploy cluster services

bert-e commented:
# Hello nicolast,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/bootstrap-orchestrate` into
`development/2.0`
* [w/2.1/improvement/bootstrap-orchestrate](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/bootstrap-orchestrate) will be merged into `development/2.1` 
* [w/2.2/improvement/bootstrap-orchestrate](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/bootstrap-orchestrate) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

NicolasT commented:
@bert-e wait

TeddyAndrieux commented on salt/metalk8s/orchestrate/bootstrap/init.sls:
```diff
@@ -0,0 +1,141 @@
+Sync bootstrap minion:
+  salt.function:
+  - name: saltutil.sync_all
+  - tgt: {{ pillar.bootstrap_id }}
```
You need to specify `saltenv` here

NicolasT replied:
You're right, done.

TeddyAndrieux commented on salt/metalk8s/orchestrate/bootstrap/init.sls:
```diff
@@ -0,0 +1,141 @@
+  - name: saltutil.refresh_pillar
+  - tgt: {{ pillar.bootstrap_id }}
+  - require:
+    - salt: Configure bootstrap Node object
```
This require is a left over I think ?

NicolasT replied:
No. The Pillar `top` is derived from the node/minion roles, and these are derived from annotations on the `Node` object, which needs to be configured correctly before we can continue.

TeddyAndrieux replied:
Yes indeed just when I commented the step "salt.runner Configure bootstrap Node object" just above was not here :stuck_out_tongue:

TeddyAndrieux reviewed (commented):
Approach LGTM

NicolasT committed:
salt: fix permissions on SA public key

This fixes idempotence between the `ca` role creating this keypair, and
the `master` role consuming it.

NicolasT committed:
salt/ui: retrieve SaltAPI IP from Pillar

NicolasT committed:
salt/kubernetes: introduce `infra` role

NicolasT committed:
salt: rework bootstrap process and roles

This patch includes the following changes:

- Define all roles in a `README` document
- Rework all roles according to the roles defined in said document
- Rework the steps from the `bootstrap.sh` script into a Salt
  orchestration SLS, to the extent possible
- Remove the call to `crictl exec ... salt-key -Ay` to accept the
  bootstrap node minion key in favor of a custom orchestration step

See: #1027
See: https://github.com/scality/metalk8s/pull/1027
See: #925
See: https://github.com/scality/metalk8s/issues/925

bert-e commented:
# History mismatch



Merge commit #05e6b4017d1eb1cfa25fa0a2101732322be1b9c6 on the integration branch
`w/2.1/improvement/bootstrap-orchestrate` is merging a branch which is neither the current
branch `improvement/bootstrap-orchestrate` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/bootstrap-orchestrate` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

NicolasT commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/bootstrap-orchestrate` into
`development/2.0`
* [w/2.1/improvement/bootstrap-orchestrate](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/bootstrap-orchestrate) will be merged into `development/2.1` 
* [w/2.2/improvement/bootstrap-orchestrate](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/bootstrap-orchestrate) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

NicolasT commented:
@bert-e approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf










*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #ad208e5e5ff21a8ccaa810e94b61930821c6ac93 on the integration branch
`w/2.1/improvement/bootstrap-orchestrate` is merging a branch which is neither the current
branch `improvement/bootstrap-orchestrate` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/bootstrap-orchestrate` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

NicolasT commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/bootstrap-orchestrate` into
`development/2.0`
* [w/2.1/improvement/bootstrap-orchestrate](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/bootstrap-orchestrate) will be merged into `development/2.1` 
* [w/2.2/improvement/bootstrap-orchestrate](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/bootstrap-orchestrate) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf










*The following options are set:* **approve**

TeddyAndrieux reviewed (approved):
LGTM

NicolasT commented:
@bert-e status

bert-e commented:
# Status





*Status report is not available.*






*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye nicolast.
